kymatio/scattering1d/backend/numpy_backend.py

```diff
@@ -0,0 +1,213 @@
+        PyTorch API for other options.  Defaults to 'constant'.
+    value : float, optional
+        If mode == 'constant', value to input within the padding. Defaults to
+        0.
+
+    Returns
+    -------
+    res : tensor
+        The tensor passed along the third dimension.
+    """
+    if (pad_left >= x.shape[-1]) or (pad_right >= x.shape[-1]):
+        if mode == 'reflect':
+            raise ValueError('Indefinite padding size (larger than tensor).')
+    return np.pad(x, ((0,0), (0,0), (pad_left, pad_right)), mode='reflect')
+
+def pad(x, pad_left=0, pad_right=0, to_complex=True):
+    """Pad real 1D tensors and map to complex
+
+    Padding which allows to simultaneously pad in a reflection fashion and map
+    to complex if necessary.
+
+    Parameters
+    ----------
+    x : tensor
+        Three-dimensional input tensor with the third axis being the one to
+        be padded.
+    pad_left : int
+        Amount to add on the left of the tensor (at the beginning of the
+        temporal axis).
+    pad_right : int
+        amount to add on the right of the tensor (at the end of the temporal
+        axis).
+    to_complex : boolean, optional
+        Whether to map the resulting padded tensor to a complex type (seen
+        as a real number). Defaults to True.
+
+    Returns
+    -------
+    output : tensor
+        A padded signal, possibly transformed into a four-dimensional tensor
+        with the last axis of size 2 if to_complex is True (this axis
+        corresponds to the real and imaginary parts).
+    """
+    output = pad_1d(x, pad_left, pad_right, mode='reflect')
+    return output
+
+def unpad(x, i0, i1):
+    """Unpad real 1D tensor
+
+    Slices the input tensor at indices between i0 and i1 along the last axis.
+
+    Parameters
+    ----------
+    x : tensor
+        Input tensor with least one axis.
+    i0 : int
+        Start of original signal before padding.
+    i1 : int
+        End of original signal before padding.
+
+    Returns
+    -------
+    x_unpadded : tensor
+        The tensor x[..., i0:i1].
+    """
+    return x[..., i0:i1]
+
+def real(x):
+    """Real part of complex tensor
+
+    Takes the real part of a complex tensor, where the last axis corresponds
+    to the real and imaginary parts.
+
+    Parameters
+    ----------
+    x : tensor
+        A complex tensor (that is, whose last dimension is equal to 2).
+
+    Returns
+    -------
+    x_real : tensor
+        The tensor x[..., 0] which is interpreted as the real part of x.
+    """
+    return np.real(x)
+
+def fft1d_c2c(x):
+    """Compute the 1D FFT of a complex signal
+
+    Input
+    -----
+    x : tensor
+        A tensor of size (..., T, 2), where x[..., 0] is the real part and
+        x[..., 1] is the imaginary part.
+
+    Returns
+    -------
+    x_f : tensor
+        A tensor of the same size as x containing its Fourier transform in the
+        standard PyTorch FFT ordering.
+    """
+    return np.fft.fft(x)
```

Comment:
I agree that it seems to be asking for trouble, but I question if the effort to refactor the numpy and tf backends are worth it if we consider that the PyTorch developers are in the process of creating a complex datatype as seen over [here](https://github.com/pytorch/pytorch/issues/755).

The discussion sounds like the datatype may not be added for a while, but when the complex datatype is implemented I question if it would be worth the cost to refactor the codebase to support the complex datatype across all backends. They claim over [here](https://github.com/pytorch/pytorch/issues/1033) that there isn't much community need for complex tensors. I believe that it would be beneficial to add our voice to the conversation, to let them know that there is a need. 

From my point of view there are two ways we could proceed: 

We could refactor the tf and np backends similar to how the torch backend does it, complex tensors/numpy arrays as float64 with a final dim of 2, and correct the documentation. (I would be happy to do this, it would not be bad). When torch releases with the complex datatype, there could be a discussion on if it makes sense to integrate the datatype into the torch backend. My guess is that it would make sense, at which point we would want to refactor the tf and np backends. This seems like more overall amount of work, but we don't know when torch releases its implementation of the complex tensor. So the design/implementation would be the same across all backends for a while. 

Alternatively, we could keep the backends as is, and mention somewhere in the documentation on kymat.io the major differences between the torch backend and the tf/np backends. When the complex tensor datatype release for Torch, we have only one backend to change across *d. This seems like less overall amount of work, but we would have two different sets of designs until the release.

Thoughts @edouardoyallon , @janden , @eickenberg ?

Reply:
> complex128, whereas torch outputs an array with shape x.shape + (2,) and datatype float64

I think a strength of the backend is to be insensitive to this. Basically, the way a type is implemented is a low-level consideration and I think it should remain as it is. As mentionned by @MuawizChaudhary , complex tensor might arrive in torch that would allow us to completely get rid of this extra 2 dimension in torch.

Reply:
Thus I'm pros keeping it as it is!

Reply:
Ok we can do that. This means that we have to be extra careful when working with different backends, however. More annoying is the fact that the docs for the different backends will no longer be identical (we can't just copy and paste them between one another) so we'll have to maintain them separately.